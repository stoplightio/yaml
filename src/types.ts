import { IParserResult } from '@stoplight/types';
import { LoadOptions, YAMLNode } from 'yaml-ast-parser';

export interface IParseOptions extends LoadOptions {
  mergeKeys?: boolean;
}

export type YamlParserResult<T> = IParserResult<T, YAMLNode, number[], IParseOptions>;
