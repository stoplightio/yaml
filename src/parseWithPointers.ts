import { DiagnosticSeverity, IDiagnostic, Optional } from '@stoplight/types';
import {
  determineScalarType,
  load as loadAST,
  parseYamlBoolean,
  parseYamlFloat,
  parseYamlInteger,
  YAMLException,
} from '@stoplight/yaml-ast-parser';
import { buildJsonPath } from './buildJsonPath';
import { SpecialMappingKeys } from './consts';
import { dereferenceAnchor } from './dereferenceAnchor';
import { lineForPosition } from './lineForPosition';
import { IParseOptions, Kind, ScalarType, YAMLMapping, YAMLNode, YamlParserResult, YAMLScalar } from './types';
import { isObject } from './utils';

export const parseWithPointers = <T>(value: string, options?: IParseOptions): YamlParserResult<T | undefined> => {
  const lineMap = computeLineMap(value);
  const ast = loadAST(value, {
    ...options,
    ignoreDuplicateKeys: true,
  }) as YAMLNode;

  const parsed: YamlParserResult<T | undefined> = {
    ast,
    lineMap,
    data: undefined,
    diagnostics: [],
    metadata: options,
  };

  if (!ast) return parsed;

  parsed.data = walkAST(ast, options, lineMap, parsed.diagnostics) as T;

  if (ast.errors) {
    parsed.diagnostics.push(...transformErrors(ast.errors, lineMap));
  }

  if (parsed.diagnostics.length > 0) {
    parsed.diagnostics.sort((itemA, itemB) => itemA.range.start.line - itemB.range.start.line);
  }

  if (Array.isArray(parsed.ast.errors)) {
    parsed.ast.errors.length = 0;
  }

  return parsed;
};

const KEYS = Symbol('object_keys');

export const walkAST = (
  node: YAMLNode | null,
  options: Optional<IParseOptions>,
  lineMap: number[],
  diagnostics: IDiagnostic[],
): unknown => {
  if (node) {
    switch (node.kind) {
      case Kind.MAP: {
        const preserveKeyOrder = options !== void 0 && options.preserveKeyOrder === true;
        const container = createMapContainer(preserveKeyOrder);
        // note, we don't handle null aka '~' keys on purpose
        const seenKeys: string[] = [];
        const handleMergeKeys = options !== void 0 && options.mergeKeys === true;
        const yamlMode = options !== void 0 && options.json === false;
        const handleDuplicates = options !== void 0 && options.ignoreDuplicateKeys === false;

        for (const mapping of node.mappings) {
          if (!validateMappingKey(mapping, lineMap, diagnostics, yamlMode)) continue;

          const key = String(getScalarValue(mapping.key));

          if ((yamlMode || handleDuplicates) && (!handleMergeKeys || key !== SpecialMappingKeys.MergeKey)) {
            if (seenKeys.includes(key)) {
              if (yamlMode) {
                throw new Error('Duplicate YAML mapping key encountered');
              }

              if (handleDuplicates) {
                diagnostics.push(createYAMLException(mapping.key, lineMap, 'duplicate key'));
              }
            } else {
              seenKeys.push(key);
            }
          }

          // https://yaml.org/type/merge.html merge keys, not a part of YAML spec
          if (handleMergeKeys && key === SpecialMappingKeys.MergeKey) {
            const reduced = reduceMergeKeys(walkAST(mapping.value, options, lineMap, diagnostics), preserveKeyOrder);
            if (preserveKeyOrder && reduced !== null) {
              for (const reducedKey of Object.keys(reduced)) {
                pushKey(container, reducedKey);
              }
            }

            Object.assign(container, reduced);
          } else {
            container[key] = walkAST(mapping.value, options, lineMap, diagnostics);

            if (preserveKeyOrder) {
              pushKey(container, key);
            }
          }
        }

        if (KEYS in container) {
          (container as Partial<{ [KEYS]: Array<Symbol | string> }>)[KEYS]!.push(KEYS);
        }

        return container;
      }
      case Kind.SEQ:
        return node.items.map(item => walkAST(item, options, lineMap, diagnostics));
      case Kind.SCALAR:
        return getScalarValue(node);
      case Kind.ANCHOR_REF: {
        if (isObject(node.value)) {
          node.value = dereferenceAnchor(node.value, node.referencesAnchor)!;
        }

        return walkAST(node.value!, options, lineMap, diagnostics);
      }
      default:
        return null;
    }
  }

  return node;
};

function getScalarValue(node: YAMLScalar): number | null | boolean | string | void {
  switch (determineScalarType(node)) {
    case ScalarType.null:
      return null;
    case ScalarType.string:
      return String(node.value);
    case ScalarType.bool:
      return parseYamlBoolean(node.value);
    case ScalarType.int:
      return parseYamlInteger(node.value);
    case ScalarType.float:
      return parseYamlFloat(node.value);
  }
}

// builds up the line map, for use by linesForPosition
const computeLineMap = (input: string) => {
  const lineMap: number[] = [];

  let i = 0;
  for (; i < input.length; i++) {
    if (input[i] === '\n') {
      lineMap.push(i + 1);
    }
  }

  lineMap.push(i + 1);

  return lineMap;
};

function getLineLength(lineMap: number[], line: number) {
  if (line === 0) {
    return Math.max(0, lineMap[0] - 1);
  }

  return Math.max(0, lineMap[line] - lineMap[line - 1] - 1);
}

const transformErrors = (errors: YAMLException[], lineMap: number[]): IDiagnostic[] => {
  const validations: IDiagnostic[] = [];
  for (const error of errors) {
    const validation: IDiagnostic = {
      code: error.name,
      message: error.reason,
      severity: error.isWarning ? DiagnosticSeverity.Warning : DiagnosticSeverity.Error,
      range: {
        start: {
          line: error.mark.line,
          character: error.mark.column,
        },
        end: {
          line: error.mark.line,
          character: error.mark.toLineEnd ? getLineLength(lineMap, error.mark.line) : error.mark.column,
        },
      },
    };

    validations.push(validation);
  }

  return validations;
};

const reduceMergeKeys = (items: unknown, preserveKeyOrder: boolean): object | null => {
  if (Array.isArray(items)) {
    // reduceRight is on purpose here! We need to respect the order - the key cannot be overridden
    const reduced = items.reduceRight(
      preserveKeyOrder
        ? (merged, item) => {
            const keys = Object.keys(item);
            for (let i = keys.length - 1; i >= 0; i--) {
              unshiftKey(merged, keys[i]);
            }

            return Object.assign(merged, item);
          }
        : (merged, item) => Object.assign(merged, item),
      createMapContainer(preserveKeyOrder),
    );

    if (preserveKeyOrder) {
      reduced[KEYS].push(KEYS);
    }

    return reduced;
  }

  return typeof items !== 'object' || items === null ? null : Object(items);
};

const traps = {
  ownKeys(target: object) {
    return target[KEYS];
  },
};

function createMapContainer(preserveKeyOrder: boolean): { [key in PropertyKey]: unknown } {
  if (preserveKeyOrder) {
    const container = new Proxy({}, traps);
    Reflect.defineProperty(container, KEYS, {
      value: [],
    });

    return container;
  }

  return {};
}

function deleteKey(container: object, key: string) {
  const index = key in container ? container[KEYS].indexOf(key) : -1;
  if (index !== -1) {
    container[KEYS].splice(index, 1);
  }
}

function unshiftKey(container: object, key: string) {
  deleteKey(container, key);
  container[KEYS].unshift(key);
}

function pushKey(container: object, key: string) {
  deleteKey(container, key);
  container[KEYS].push(key);
}

function validateMappingKey(
  mapping: YAMLMapping,
  lineMap: number[],
  diagnostics: IDiagnostic[],
  yamlMode: boolean,
): boolean {
  if (mapping.key.kind !== Kind.SCALAR) {
    if (!yamlMode) {
      diagnostics.push(
        createYAMLIncompatibilityException(mapping.key, lineMap, 'mapping key must be a string scalar', yamlMode),
      );
    }

    // no exception is thrown, yet the mapping is excluded regardless of mode, as we cannot represent the value anyway
    return false;
  }

  if (!yamlMode) {
    const type = typeof getScalarValue(mapping.key);
    if (type !== 'string') {
      diagnostics.push(
        createYAMLIncompatibilityException(
          mapping.key,
          lineMap,
          `mapping key must be a string scalar rather than ${mapping.key.valueObject === null ? 'null' : type}`,
          yamlMode,
        ),
      );
    }
  }

  return true;
}

function createYAMLIncompatibilityException(
  node: YAMLNode,
  lineMap: number[],
  message: string,
  yamlMode: boolean,
): IDiagnostic {
  const exception = createYAMLException(node, lineMap, message);
  exception.code = 'YAMLIncompatibleValue';
  exception.severity = yamlMode ? DiagnosticSeverity.Hint : DiagnosticSeverity.Warning;
  return exception;
}

function createYAMLException(node: YAMLNode, lineMap: number[], message: string): IDiagnostic {
  const startLine = lineForPosition(node.startPosition, lineMap);
  const endLine = lineForPosition(node.endPosition, lineMap);

  return {
    code: 'YAMLException',
    message,
    severity: DiagnosticSeverity.Error,
    path: buildJsonPath(node),
    range: {
      start: {
        line: startLine,
        character: startLine === 0 ? node.startPosition : node.startPosition - lineMap[startLine - 1],
      },
      end: {
        line: endLine,
        character: endLine === 0 ? node.endPosition : node.endPosition - lineMap[endLine - 1],
      },
    },
  };
}
