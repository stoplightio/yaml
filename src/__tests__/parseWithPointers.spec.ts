import { DiagnosticSeverity } from '@stoplight/types';
import * as fs from 'fs';
import * as path from 'path';
import { parseWithPointers } from '../parseWithPointers';
import * as HugeJSON from './fixtures/huge-json.json';
import { HugeYAML } from './fixtures/huge-yaml';

const diverse = fs.readFileSync(path.join(__dirname, './fixtures/diverse.yaml'), 'utf-8');

describe('yaml parser', () => {
  test.each(['test', 1])('parse scalar $s', val => {
    const result = parseWithPointers(String(val));
    expect(result.data).toEqual(val);
  });

  test('parse sequences', () => {
    const result = parseWithPointers('[0, 1, 2]');
    expect(result.data).toEqual([0, 1, 2]);
  });

  test('parse diverse', () => {
    const result = parseWithPointers(diverse);
    expect(result).toMatchSnapshot();
  });

  test('parse huge', () => {
    const result = parseWithPointers(HugeYAML);
    expect(result.data).toEqual(HugeJSON);
  });

  describe('report errors', () => {
    test('unknown tags', () => {
      const result = parseWithPointers(`function: !!js/function >
  function foobar() {
    return 'Wow! JS-YAML Rocks!';
  }
  
test: !!css >
  function boom() {}
`);

      expect(result).toHaveProperty('diagnostics', [
        {
          code: 'YAMLException',
          message: 'unknown tag <tag:yaml.org,2002:js/function>',
          range: {
            end: {
              character: 25,
              line: 0,
            },
            start: {
              character: 10,
              line: 0,
            },
          },
          severity: DiagnosticSeverity.Error,
        },
        {
          code: 'YAMLException',
          message: 'unknown tag <tag:yaml.org,2002:css>',
          range: {
            end: {
              character: 13,
              line: 5,
            },
            start: {
              character: 6,
              line: 5,
            },
          },
          severity: DiagnosticSeverity.Error,
        },
      ]);
    });

    test('invalid mapping', () => {
      const result = parseWithPointers(
        `prop1: true
prop2: true
  inner 1
  val: 2`
      );

      expect(result.diagnostics).toEqual([
        {
          severity: DiagnosticSeverity.Error,
          message: 'bad indentation of a mapping entry',
          code: 'YAMLException',
          range: {
            start: {
              character: 5,
              line: 3,
            },
            end: {
              character: 5,
              line: 3,
            },
          },
        },
        {
          severity: DiagnosticSeverity.Error,
          message: 'incomplete explicit mapping pair; a key node is missed',
          code: 'YAMLException',
          range: {
            start: {
              character: 7,
              line: 3,
            },
            end: {
              character: 7,
              line: 3,
            },
          },
        },
      ]);
    });
  });

  describe('dereferencing anchor refs', () => {
    test('ignore valid refs', () => {
      const result = parseWithPointers(`austrian-cities: &austrian-cities
  - Vienna
  - Graz
  - Linz
  - Salzburg
  
european-cities:
  austria: *austrian-cities
`);
      expect(result.data).toEqual({
        'austrian-cities': ['Vienna', 'Graz', 'Linz', 'Salzburg'],
        'european-cities': {
          austria: ['Vienna', 'Graz', 'Linz', 'Salzburg'],
        },
      });
    });

    test('support circular refs in mapping', () => {
      const result = parseWithPointers(`definitions:
  model: &ref
    foo:
      name: *ref
`);

      expect(result.data).toEqual({
        definitions: {
          model: {
            foo: {
              name: {
                foo: {
                  name: undefined,
                },
              },
            },
          },
        },
      });

      expect(() => JSON.stringify(result.data)).not.toThrow();
    });

    test('support circular in refs sequences', () => {
      const result = parseWithPointers(`- &foo
  - test:
    - *foo
`);
      expect(result.data).toEqual([
        [
          {
            test: [[{ test: [undefined] }]],
          },
        ],
      ]);

      expect(() => JSON.stringify(result.data)).not.toThrow();
    });

    test('support mixed refs', () => {
      const result = parseWithPointers(`austrian-cities: &austrian-cities
  - Vienna
  - Graz
  - Linz
  - Salzburg
  
european-cities: &cities
    austria: *austrian-cities
    all: *cities
`);

      expect(result.data).toEqual({
        'austrian-cities': ['Vienna', 'Graz', 'Linz', 'Salzburg'],
        'european-cities': {
          all: {
            austria: ['Vienna', 'Graz', 'Linz', 'Salzburg'],
          },
          austria: ['Vienna', 'Graz', 'Linz', 'Salzburg'],
        },
      });

      expect(() => JSON.stringify(result.data)).not.toThrow();
    });

    test('support circular nested refs', () => {
      const result = parseWithPointers(`a: &foo
  - b: &bar
    - true
    - c: *bar
    - *foo
`);

      expect(result.data).toEqual({
        a: [
          {
            b: [true, { c: [true, { c: undefined }, undefined] }, [{ b: [true, { c: undefined }, undefined] }]],
          },
        ],
      });

      expect(() => JSON.stringify(result.data)).not.toThrow();
    });

    test('insane edge case', () => {
      const result = parseWithPointers(`- &foo
  - *foo
  - test:
    - *foo
  - abc: &test
      foo: 2
      a:
      c: *foo
      x: *test
`);
      expect(result.data).toMatchSnapshot();

      expect(() => JSON.stringify(result.data)).not.toThrow();
    });
  });
});
