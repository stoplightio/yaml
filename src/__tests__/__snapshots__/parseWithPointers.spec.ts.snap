// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yaml parser dereferencing anchor refs insane edge case 1`] = `
Array [
  Array [
    Array [
      undefined,
      Object {
        "test": Array [
          undefined,
        ],
      },
      Object {
        "abc": Object {
          "a": null,
          "c": undefined,
          "foo": 2,
          "x": undefined,
        },
      },
    ],
    Object {
      "test": Array [
        Array [
          undefined,
          Object {
            "test": Array [
              undefined,
            ],
          },
          Object {
            "abc": Object {
              "a": null,
              "c": undefined,
              "foo": 2,
              "x": undefined,
            },
          },
        ],
      ],
    },
    Object {
      "abc": Object {
        "a": null,
        "c": Array [
          undefined,
          Object {
            "test": Array [
              undefined,
            ],
          },
          Object {
            "abc": Object {
              "a": null,
              "c": undefined,
              "foo": 2,
              "x": undefined,
            },
          },
        ],
        "foo": 2,
        "x": Object {
          "a": null,
          "c": undefined,
          "foo": 2,
          "x": undefined,
        },
      },
    },
  ],
]
`;

exports[`yaml parser parse diverse 1`] = `
Object {
  "ast": Object {
    "endPosition": 442,
    "errors": Array [],
    "kind": 2,
    "mappings": Array [
      Object {
        "endPosition": 189,
        "errors": Array [],
        "key": Object {
          "doubleQuoted": false,
          "endPosition": 145,
          "errors": Array [],
          "kind": 0,
          "parent": [Circular],
          "plainScalar": true,
          "rawValue": "json",
          "startPosition": 141,
          "value": "json",
        },
        "kind": 1,
        "parent": [Circular],
        "startPosition": 141,
        "value": Object {
          "endPosition": 189,
          "errors": Array [],
          "items": Array [
            Object {
              "doubleQuoted": false,
              "endPosition": 156,
              "errors": Array [],
              "kind": 0,
              "parent": [Circular],
              "plainScalar": true,
              "rawValue": "rigid",
              "startPosition": 151,
              "value": "rigid",
            },
            Object {
              "doubleQuoted": false,
              "endPosition": 188,
              "errors": Array [],
              "kind": 0,
              "parent": [Circular],
              "plainScalar": true,
              "rawValue": "better for data interchange",
              "startPosition": 161,
              "value": "better for data interchange",
            },
          ],
          "kind": 3,
          "parent": [Circular],
          "startPosition": 149,
        },
      },
      Object {
        "endPosition": 234,
        "errors": Array [],
        "key": Object {
          "doubleQuoted": false,
          "endPosition": 193,
          "errors": Array [],
          "kind": 0,
          "parent": [Circular],
          "plainScalar": true,
          "rawValue": "yaml",
          "startPosition": 189,
          "value": "yaml",
        },
        "kind": 1,
        "parent": [Circular],
        "startPosition": 189,
        "value": Object {
          "endPosition": 234,
          "errors": Array [],
          "items": Array [
            Object {
              "doubleQuoted": false,
              "endPosition": 216,
              "errors": Array [],
              "kind": 0,
              "parent": [Circular],
              "plainScalar": true,
              "rawValue": "slim and flexible",
              "startPosition": 199,
              "value": "slim and flexible",
            },
            Object {
              "doubleQuoted": false,
              "endPosition": 224,
              "errors": Array [],
              "kind": 0,
              "parent": [Circular],
              "plainScalar": true,
              "rawValue": "123",
              "startPosition": 221,
              "value": "123",
              "valueObject": 123,
            },
            Object {
              "doubleQuoted": false,
              "endPosition": 233,
              "errors": Array [],
              "kind": 0,
              "parent": [Circular],
              "plainScalar": true,
              "rawValue": "true",
              "startPosition": 229,
              "value": "true",
              "valueObject": true,
            },
          ],
          "kind": 3,
          "parent": [Circular],
          "startPosition": 197,
        },
      },
      Object {
        "endPosition": 319,
        "errors": Array [],
        "key": Object {
          "doubleQuoted": false,
          "endPosition": 240,
          "errors": Array [],
          "kind": 0,
          "parent": [Circular],
          "plainScalar": true,
          "rawValue": "object",
          "startPosition": 234,
          "value": "object",
        },
        "kind": 1,
        "parent": [Circular],
        "startPosition": 234,
        "value": Object {
          "endPosition": 319,
          "errors": Array [],
          "kind": 2,
          "mappings": Array [
            Object {
              "endPosition": 254,
              "errors": Array [],
              "key": Object {
                "doubleQuoted": false,
                "endPosition": 247,
                "errors": Array [],
                "kind": 0,
                "parent": [Circular],
                "plainScalar": true,
                "rawValue": "key",
                "startPosition": 244,
                "value": "key",
              },
              "kind": 1,
              "parent": [Circular],
              "startPosition": 244,
              "value": Object {
                "doubleQuoted": false,
                "endPosition": 254,
                "errors": Array [],
                "kind": 0,
                "parent": [Circular],
                "plainScalar": true,
                "rawValue": "value",
                "startPosition": 249,
                "value": "value",
              },
            },
            Object {
              "endPosition": 319,
              "errors": Array [],
              "key": Object {
                "doubleQuoted": false,
                "endPosition": 262,
                "errors": Array [],
                "kind": 0,
                "parent": [Circular],
                "plainScalar": true,
                "rawValue": "array",
                "startPosition": 257,
                "value": "array",
              },
              "kind": 1,
              "parent": [Circular],
              "startPosition": 257,
              "value": Object {
                "endPosition": 319,
                "errors": Array [],
                "items": Array [
                  Object {
                    "endPosition": 281,
                    "errors": Array [],
                    "kind": 2,
                    "mappings": Array [
                      Object {
                        "endPosition": 281,
                        "errors": Array [],
                        "key": Object {
                          "doubleQuoted": false,
                          "endPosition": 280,
                          "errors": Array [],
                          "kind": 0,
                          "parent": [Circular],
                          "plainScalar": true,
                          "rawValue": "null_value",
                          "startPosition": 270,
                          "value": "null_value",
                        },
                        "kind": 1,
                        "parent": [Circular],
                        "startPosition": 270,
                        "value": null,
                      },
                    ],
                    "parent": [Circular],
                    "startPosition": 270,
                  },
                  Object {
                    "endPosition": 301,
                    "errors": Array [],
                    "kind": 2,
                    "mappings": Array [
                      Object {
                        "endPosition": 301,
                        "errors": Array [],
                        "key": Object {
                          "doubleQuoted": false,
                          "endPosition": 295,
                          "errors": Array [],
                          "kind": 0,
                          "parent": [Circular],
                          "plainScalar": true,
                          "rawValue": "boolean",
                          "startPosition": 288,
                          "value": "boolean",
                        },
                        "kind": 1,
                        "parent": [Circular],
                        "startPosition": 288,
                        "value": Object {
                          "doubleQuoted": false,
                          "endPosition": 301,
                          "errors": Array [],
                          "kind": 0,
                          "parent": [Circular],
                          "plainScalar": true,
                          "rawValue": "true",
                          "startPosition": 297,
                          "value": "true",
                          "valueObject": true,
                        },
                      },
                    ],
                    "parent": [Circular],
                    "startPosition": 288,
                  },
                  Object {
                    "endPosition": 318,
                    "errors": Array [],
                    "kind": 2,
                    "mappings": Array [
                      Object {
                        "endPosition": 318,
                        "errors": Array [],
                        "key": Object {
                          "doubleQuoted": false,
                          "endPosition": 315,
                          "errors": Array [],
                          "kind": 0,
                          "parent": [Circular],
                          "plainScalar": true,
                          "rawValue": "integer",
                          "startPosition": 308,
                          "value": "integer",
                        },
                        "kind": 1,
                        "parent": [Circular],
                        "startPosition": 308,
                        "value": Object {
                          "doubleQuoted": false,
                          "endPosition": 318,
                          "errors": Array [],
                          "kind": 0,
                          "parent": [Circular],
                          "plainScalar": true,
                          "rawValue": "1",
                          "startPosition": 317,
                          "value": "1",
                          "valueObject": 1,
                        },
                      },
                    ],
                    "parent": [Circular],
                    "startPosition": 308,
                  },
                ],
                "kind": 3,
                "parent": [Circular],
                "startPosition": 268,
              },
            },
          ],
          "parent": [Circular],
          "startPosition": 244,
        },
      },
      Object {
        "endPosition": 372,
        "errors": Array [],
        "key": Object {
          "doubleQuoted": false,
          "endPosition": 328,
          "errors": Array [],
          "kind": 0,
          "parent": [Circular],
          "plainScalar": true,
          "rawValue": "paragraph",
          "startPosition": 319,
          "value": "paragraph",
        },
        "kind": 1,
        "parent": [Circular],
        "startPosition": 319,
        "value": Object {
          "doubleQuoted": false,
          "endPosition": 372,
          "errors": Array [],
          "kind": 0,
          "parent": [Circular],
          "rawValue": ">
  Blank lines denote

  paragraph breaks",
          "startPosition": 330,
          "value": "Blank lines denote
paragraph breaks
",
        },
      },
      Object {
        "endPosition": 441,
        "errors": Array [],
        "key": Object {
          "doubleQuoted": false,
          "endPosition": 380,
          "errors": Array [],
          "kind": 0,
          "parent": [Circular],
          "plainScalar": true,
          "rawValue": "content",
          "startPosition": 373,
          "value": "content",
        },
        "kind": 1,
        "parent": [Circular],
        "startPosition": 373,
        "value": Object {
          "doubleQuoted": false,
          "endPosition": 441,
          "errors": Array [],
          "kind": 0,
          "parent": [Circular],
          "rawValue": "|-
  Or we
  can auto
  convert line breaks
  to save space",
          "startPosition": 382,
          "value": "Or we
can auto
convert line breaks
to save space",
        },
      },
    ],
    "parent": null,
    "startPosition": 141,
  },
  "data": Object {
    "content": "Or we
can auto
convert line breaks
to save space",
    "json": Array [
      "rigid",
      "better for data interchange",
    ],
    "object": Object {
      "array": Array [
        Object {
          "null_value": null,
        },
        Object {
          "boolean": true,
        },
        Object {
          "integer": 1,
        },
      ],
      "key": "value",
    },
    "paragraph": "Blank lines denote
paragraph breaks
",
    "yaml": Array [
      "slim and flexible",
      123,
      true,
    ],
  },
  "diagnostics": Array [],
  "lineMap": Array [
    4,
    47,
    90,
    123,
    140,
    141,
    147,
    157,
    189,
    195,
    217,
    225,
    234,
    242,
    255,
    264,
    282,
    302,
    319,
    332,
    353,
    354,
    373,
    385,
    393,
    404,
    426,
    442,
    443,
  ],
  "metadata": undefined,
}
`;
